### PR TITLE
FATES parteh scale fixes api (ie nutrient + methane coupling)

### PR DESCRIPTION
**Pull request by rgknox**
### Description of changes

This set of changes is necessary to have API compatibility with FATES PR: https://github.com/NGEET/fates/pull/716
In brief, these changes accomodate boundary conditions needed for nutrient coupling, although CLM-FATES nutrient coupling is forthcoming.  These changes are also needed to enable methane coupling in with FATES-CLM which should be enabled now.

### Specific notes

Tested this on cheyenne with modifications to check B4B, which generated expected results.
Here are the sub-branches to this PR that have the non-b4b changes suspended:

https://github.com/rgknox/fates/tree/parteh-scaling-fixes-may05-b4b
https://github.com/rgknox/ctsm/tree/fates_parteh_scale_fixes_api_b4b

Contributors other than yourself, if any: see https://github.com/NGEET/fates/pull/716

CTSM Issues Fixed (include github issue #):

Are answers expected to change (and if so in what way)?

yes

Any User Interface Changes (namelist or namelist defaults changes)?

A new default parameter file has been loaded.  
No new NL entries.
No changes in default NL entries.
use_fates and use_lch4 can now both be true.

Testing performed, if any:

The FATES test suite was run on the B4B sub-branches, results pass and are found here:

 /glade/scratch/rgknox/clmed-tests/parteh-scaling-fixes-api-C7bd490c2-Fabb43f0b-test-b4b-v4.fates.cheyenne

(List what testing you did to show your changes worked as expected)
(This can be manual testing or running of the different test suites)
(Documentation on system testing is here: https://github.com/ESCOMP/ctsm/wiki/System-Testing-Guide)
(aux_clm on cheyenne for intel/gnu and izumi for intel/gnu/nag/pgi is the standard for tags on master)

**NOTE: Be sure to check your coding style against the standard
(https://github.com/ESCOMP/ctsm/wiki/CTSM-coding-guidelines) and review
the list of common problems to watch out for
(https://github.com/ESCOMP/CTSM/wiki/List-of-common-problems).**

### CODE REVIEW

**Comment by rgknox** [comment]
Imported new parameter file:
```
cseg/inputdata> ./rimport -file lnd/clm2/paramdata/fates_params_api.15.1.0_12pft_c210505.nc
 
svn import  /glade/p/cesm/cseg/inputdata/lnd/clm2/paramdata/fates_params_api.15.1.0_12pft_c210505.nc https://svn-ccsm-inputdata.cgd.ucar.edu/trunk/inputdata/lnd/clm2/paramdata/fates_params_api.15.1.0_12pft_c210505.nc
Adding  (bin)  /glade/p/cesm/cseg/inputdata/lnd/clm2/paramdata/fates_params_api.15.1.0_12pft_c210505.nc
```

**Comment by rgknox** [comment]
Addressed some issues brought up in the ctsm software meeting:
1) the new fates-ch4 test is now based off of FatesColdDef
2) the new fates-ch4 tests are now reduced in scope (no izumi on aux_clm) and less resource intensive (f10 grid, 9 days instead of f45 and 30 days)

File: src/biogeochem/ch4Mod.F90 [hunk 1]
```diff
@@ -2952,117 +2969,251 @@ subroutine ch4_aere (bounds, num_methc, filter_methc, num_methp, filter_methp, &
...
+      deallocate(oxaere)
+      deallocate(rootfr_vr)
+        
     end associate
 
   end subroutine ch4_aere
 
+  !--------------------------------------------------------------------------------------
+  
+  subroutine SiteOxAere(is_vegetated, &
```
Comment (ekluzek) [inline src/biogeochem/ch4Mod.F90:3057]: We should keep in mind a little bit about efficiency. Compilers *should* optimize small single point subroutines like this. But, they don't always do a good job of it. When it's in the same module as in this case -- they should be able to though. But, we want to think about this anyway. Now, another reason to do this might be to for example add some unit testing around this SiteOxAere subroutine, as well as making it more modular. I actually think these later points are the most important, but we should always be mindful of efficiency.

File: src/biogeochem/ch4Mod.F90 [hunk 2]
```diff
@@ -2952,117 +2969,251 @@ subroutine ch4_aere (bounds, num_methc, filter_methc, num_methp, filter_methp, &
          end do
       end do
 
-      diffus_aere = d_con_g(1,1)*1.e-4_r8  ! for CH4: m^2/s
-      ! This parameter is poorly constrained and should be done on a patch-specific basis...
+      allocate(rootfr_vr(nlevsoi))
```
Comment (ekluzek) [inline src/biogeochem/ch4Mod.F90:2972]: Rather than allocate and deallocate every time you do into here. It likely would be better to make these static arrays as module data (since they are small). Then you don't have to allocate and deallocate each time you go inside. You could also make them automatic arrays so you don't have to actually do the allocate and deallocate steps, but making them module data of static size is probably the best here.

**Review by ekluzek — CHANGES_REQUESTED** [review]
I have a lot of nit picky things here. Some of it I just want to hear about your thinking. And as we have already discussed I'm fine with making an initial implementation that we don't change much, but start going toward a design we agree on.

**Comment by ekluzek** [comment]
An important issue that relates to this is https://github.com/ESCOMP/CTSM/issues/1051. We want to think of this as a step going towards some of that.

**Comment by ekluzek** [comment]
After looking at the implementation. I do think that there should be some filters setup for running over FATES points versus vegetation (but NOT FATES) points. In most cases there should be two loops one loop for each. I only saw one instance where right now there might be issues with that since there's a part that they both do within the same loop.

File: src/biogeochem/ch4Mod.F90 (hunk 2 above)
Comment (rgknox) [reply src/biogeochem/ch4Mod.F90:2972]: great idea, will do

File: src/biogeochem/ch4Mod.F90 (hunk 1 above)
Comment (rgknox) [reply src/biogeochem/ch4Mod.F90:3057]: Agreed. On the fates side of the code, I'm trying to get our key subroutines that are most tied to the science (ie where the sauce is made, like leaf-level photosynthesis, etc) using primitive arguments and being calculated on a single element.  This makes the routine usable at different scales, presumably by different model scaling strategies (ie fates or big-leaf), and more readily unit testable.  However, I do see your point about efficiency. I wonder if there is a good fortran reference where the cover the performance trade-offs of encapsulating things in their own routines, vis-a-vis allocating/shuffling/copying/pointing to memory.

File: src/utils/clmfates_interfaceMod.F90 [hunk 3]
```diff
@@ -116,7 +117,7 @@ module CLMFatesInterfaceMod
    use FatesInterfaceMod     , only : set_fates_ctrlparms
 
 
-   use FatesHistoryInterfaceMod, only : fates_history_interface_type
+   use FatesHistoryInterfaceMod, only : fates_hist
```
[the remark this answers is not in the document]
Comment (rgknox) [reply src/utils/clmfates_interfaceMod.F90:120]: This is not cosmetic actually.  We are trying to decentralize the updating of history variables on the FATES side of the code.  In order to access those history variable structures at any location, we need to do either 1 of three things:
1) pass in clm_fates%fates_hist as an argument for every call (Ryan slowly looses his mind from all the code)
2) instantiate fates_hist on its own (remove from clm_fates), and allow global access to it (this method)
3) keep clm_fates%fates_hist but also set a global pointer on the fates side of the code that points to clm_fates%fates_hist  (this is not that much different that item 2, but requires 1 more piece of information)

File: src/biogeochem/ch4Mod.F90 [hunk 4]
```diff
@@ -2952,117 +2969,251 @@ subroutine ch4_aere (bounds, num_methc, filter_methc, num_methp, filter_methp, &
...
+    real(r8) :: n_tiller 
+    real(r8) :: h2osoi_vol_min
+    real(r8) :: k_h_cc, k_h_inv
+    real(r8) :: anpp, nppratio
+    real(r8) :: conc_ch4_wat
+    real(r8) :: aerecond    ! aerenchyma conductance (m/s)
+    real(r8), parameter :: smallnumber = 1.e-12_r8
+    
+    diffus_aere = d_con_g(1,1)*1.e-4_r8  ! for CH4: m^2/s
+    ! This parameter is poorly constrained and should be done on a PFT-specific basis...
```
[the remark this answers is not in the document]
Comment (rgknox) [reply src/biogeochem/ch4Mod.F90:3127]: will move this comment to before the specification

File: src/biogeochem/ch4Mod.F90 [hunk 5]
```diff
@@ -2952,117 +2969,251 @@ subroutine ch4_aere (bounds, num_methc, filter_methc, num_methp, filter_methp, &
...
-                  ! mol/m3/s    mol/m3                                   mm / s         m           mm/m
-                  ! Use rootr here for effective per-layer transpiration, which may not be the same as rootfr
-                  tranloss = max(tranloss, 0._r8) ! in case transpiration is pathological
+         do fp = 1, num_methp
+            p = filter_methp (fp)
+            c = patch%column(p)
+            g = col%gridcell(c)
+
+            if(.not.col%is_fates(c) ) then
+               if(patch%itype(p) /= noveg) then
```
[the remark this answers is not in the document]
Comment (rgknox) [reply src/biogeochem/ch4Mod.F90:2988]: Discussed with @ekluzek, we decided to keep this logic outside SiteOxAere

File: src/main/ColumnType.F90 [hunk 6]
```diff
@@ -115,6 +118,8 @@ subroutine Init(this, begc, endc)
     allocate(this%active      (begc:endc))                     ; this%active      (:)   = .false.
     allocate(this%type_is_dynamic(begc:endc))                  ; this%type_is_dynamic(:) = .false.
 
+    allocate(this%is_fates(begc:endc))                         ; this%is_fates(:) = .false.
```
[the remark this answers is not in the document]
Comment (rgknox) [reply src/main/ColumnType.F90:121]: after discussion we decided to hold on changing this. We might change it later, but we also are not sure if we will move to make a fates filter instead of a logical mask anyway.

**Comment by rgknox** [comment]
updated test results:
cheyenne aux_clm:  /glade/scratch/rgknox/clmed-tests/parteh-scaling-fixes-api-dev039Ce8851116-Fb77a0d5d

**Comment by rgknox** [comment]
Here are the izumi tests:  /scratch/cluster/rgknox/tests_0522-093537iz

From what I can tell, results between this branch (merged with dev039) and dev039 are identical (with exceptions to fates tests) and NLCOMPs.  Although I'm not sure if the nlcomps are failing because of some write permission issues, or something else.

**Comment by rgknox** [comment]
Tests on izumi:  https://github.com/E3SM-Project/E3SM/pull/4215
From discussions with @glemieux , the nag fales with Clm50Fates _D_ are known 
```
[rgknox@izumi tests_0524-115727iz]$ ./cs.status.fails 
0524-115727iz_gnu: 4 tests

 
0524-115727iz_int: 6 tests

 
0524-115727iz_nag: 35 tests
    FAIL ERI_D_Ld9_P48x1_Vnuopc.f10_f10_mg37.I2000Clm50Sp.izumi_nag.clm-SNICARFRC SHAREDLIB_BUILD time=305
    FAIL ERS_D_Ld5.f10_f10_mg37.I2000Clm45Fates.izumi_nag.clm-FatesColdDef NLCOMP
    FAIL ERS_D_Ld5.f10_f10_mg37.I2000Clm45Fates.izumi_nag.clm-FatesColdDef RUN time=42
    FAIL ERS_D_Ld5.f10_f10_mg37.I2000Clm50Fates.izumi_nag.clm-FatesColdDef NLCOMP
    FAIL ERS_D_Ld5.f10_f10_mg37.I2000Clm50Fates.izumi_nag.clm-FatesColdDef RUN time=17
    FAIL ERS_D_Mmpi-serial_Ld5.1x1_brazil.I2000Clm45FatesRs.izumi_nag.clm-FatesColdDef NLCOMP
    FAIL ERS_D_Mmpi-serial_Ld5.1x1_brazil.I2000Clm45FatesRs.izumi_nag.clm-FatesColdDef RUN time=12
    FAIL ERS_D_Mmpi-serial_Ld5.1x1_brazil.I2000Clm50FatesRs.izumi_nag.clm-FatesColdDef NLCOMP
    FAIL ERS_D_Mmpi-serial_Ld5.1x1_brazil.I2000Clm50FatesRs.izumi_nag.clm-FatesColdDef RUN time=12
    FAIL ERS_D_Mmpi-serial_Ld5_Vnuopc.1x1_brazil.I2000Clm50FatesRs.izumi_nag.clm-FatesColdDef SHAREDLIB_BUILD time=305
    FAIL ERS_D_Mmpi-serial_Ld5_Vnuopc.1x1_brazil.I2000Clm50FatesRs.izumi_nag.clm-FatesColdDef NLCOMP
    FAIL SMS_D_Ld1_P48x1_Vnuopc.f10_f10_mg37.I2000Clm50BgcCru.izumi_nag.clm-af_bias_v7 SHAREDLIB_BUILD time=305
    FAIL SMS_D_Ld5.f10_f10_mg37.I2000Clm45Fates.izumi_nag.clm-FatesColdDef NLCOMP
    FAIL SMS_D_Ld5.f10_f10_mg37.I2000Clm45Fates.izumi_nag.clm-FatesColdDef RUN time=67
    FAIL SMS_D_Ld5.f10_f10_mg37.I2000Clm50Fates.izumi_nag.clm-FatesColdDef NLCOMP
    FAIL SMS_D_Ld5.f10_f10_mg37.I2000Clm50Fates.izumi_nag.clm-FatesColdDef RUN time=17
    FAIL SMS_D_Mmpi-serial_Ld5.5x5_amazon.I2000Clm45FatesRs.izumi_nag.clm-FatesColdDef NLCOMP
    FAIL SMS_D_Mmpi-serial_Ld5.5x5_amazon.I2000Clm45FatesRs.izumi_nag.clm-FatesColdDef RUN time=12
    FAIL SMS_D_Mmpi-serial_Ld5.5x5_amazon.I2000Clm50FatesRs.izumi_nag.clm-FatesColdDef NLCOMP
    FAIL SMS_D_Mmpi-serial_Ld5.5x5_amazon.I2000Clm50FatesRs.izumi_nag.clm-FatesColdDef RUN time=11
    FAIL SMS_D_Vnuopc_Mmpi-serial.CLM_USRDAT.I1PtClm51Bgc.izumi_nag.clm-NEON_NIWO SHAREDLIB_BUILD time=304

 
0524-115727iz_pgi: 2 tests

 
========================================================================
Non-PASS results for select phases:
TPUTCOMP non-passes: 0
MEMCOMP non-passes: 0

```

File: src/biogeochem/ch4Mod.F90 [hunk 7]
```diff
@@ -2847,36 +2878,39 @@ subroutine ch4_aere (bounds, num_methc, filter_methc, num_methp, filter_methp, &
...
-    real(r8) :: scale_factor_aere      ! scale factor on the aerenchyma area for sensitivity tests
-    real(r8) :: nongrassporosratio     ! Ratio of root porosity in non-grass to grass, used for aerenchyma transport
-    real(r8) :: unsat_aere_ratio       ! Ratio to multiply upland vegetation aerenchyma porosity by compared to inundated systems (= 0.05_r8 / 0.3_r8)
-    real(r8) :: porosmin               ! minimum aerenchyma porosity (unitless)(= 0.05_r8)
-
+    real(r8), allocatable :: tranloss(:)     ! loss due to transpiration (mol / m3 /s)
+    real(r8), allocatable :: aere(:) 
+    real(r8), allocatable :: oxaere(:)     ! (mol / m3 /s)
+    real(r8), allocatable :: rootfr_vr(:) ! Root fraction over depth
+    real(r8) :: aeretran
```
[the remark this answers is not in the document]
Comment (ekluzek) [reply src/biogeochem/ch4Mod.F90:2897]: This was a problem with the code before. And Ryan doesn't know about all the variables. So the plan was to add info if easy to find.

File: src/biogeochem/ch4Mod.F90 [hunk 8]
```diff
@@ -2952,117 +2969,251 @@ subroutine ch4_aere (bounds, num_methc, filter_methc, num_methp, filter_methp, &
...
+                    sat,              &
+                    tranloss,         & ! Out
+                    aere,             & ! Out
+                    oxaere)             ! Out  
+
+
+    use clm_varcon       , only : rpi
+    use ch4varcon        , only : transpirationloss, use_aereoxid_prog
+    
+    ! Arguments (in)
```
[the remark this answers is not in the document]
Comment (ekluzek) [reply src/biogeochem/ch4Mod.F90:3084]: Done

File: src/utils/clmfates_interfaceMod.F90 (hunk 3 above)
[the remark this answers is not in the document]
Comment (ekluzek) [reply src/utils/clmfates_interfaceMod.F90:120]: Decentralizing history is a good goal. I concur.

**Review by ekluzek — APPROVED** [review]

**Comment by ekluzek** [comment]
The failed RUN tests on izumi were due to a "dangling pointer". The nuopc tests are expected fails. @rgknox thinks he knows the solution so he's trying it out. If it looks good we'll go ahead with the tag. Otherwise, it might take more time to get it out...

**Comment by rgknox** [comment]
Updated tests on izumi and cheyenne. (merged in dev040, and performed compare):

Cheyenne:   /glade/scratch/rgknox/tests_0525-131115ch
Izumi:           /scratch/cluster/rgknox/tests_0525-125435iz

I couldn't find any tests that had unexpected fails.

Summary all pass with following expected exceptions:

Expected BASELINE AND NLCOMP fails for all fates tests.
Cheyenne:
  DAE_N2_D_Lh12_Vnuopc.f10_f10_mg37.I2000Clm50BgcCrop.cheyenne_intel.clm-DA_multidrv (expected failure)
  ERS_Ld9.f10_f10_mg37.I2000Clm50FatesCru.cheyenne_intel.clm-FatesColdDefCH4 (BFAIL - NEW TEST)

Izumi (Vnuopc fails SHAREDLIB_BUILD):
  FAIL ERI_D_Ld9_P48x1_Vnuopc.f10_f10_mg37.I2000Clm50Sp.izumi_nag.clm-SNICARFRC SHAREDLIB_BUILD 
  FAIL ERS_D_Mmpi-serial_Ld5_Vnuopc.1x1_brazil.I2000Clm50FatesRs.izumi_nag.clm-FatesColdDef SHAREDLIB_BUILD
  FAIL SMS_D_Ld1_P48x1_Vnuopc.f10_f10_mg37.I2000Clm50BgcCru.izumi_nag.clm-af_bias_v7 SHAREDLIB_BUILD
  FAIL SMS_D_Vnuopc_Mmpi-serial.CLM_USRDAT.I1PtClm51Bgc.izumi_nag.clm-NEON_NIWO SHAREDLIB_BUILD

**Comment by rgknox** [comment]
@ekluzek and I realized that the fates hash in the test, even though it was using the correct fates branch, had not merged in the most up-to-date version of fates.  So I updated the external model pointer to use the new fates tag, and re-ran tests:

cheyenne: /glade/scratch/rgknox/tests_0526-120450ch
